### PR TITLE
ticdc: Support Vector data type | tidb=master pd=master tikv=master

### PR DESCRIPTION
**Pull request by wk989898**
### What problem does this PR solve?

Issue Number: ref #11530 

### What is changed and how it works?
Detect TypeTiDBVectorFloat32, and convert vector to string in DML sink.
```sql
CREATE table test.simple(id int primary key, data VECTOR(5));
INSERT INTO test.simple(id, data) VALUES (1, '[1,2,3,4,5]');
```
Some changes like converting VECTOR to LONGTEXT on an unsupported downstream, supporting multiple output protocols may be temporary. Converting VECTOR to LONGTEXT is automatical when downstream is not TiDB or TiDB version less than 8.3.0.
```sql
CREATE table test.simple(id int primary key, data LONGTEXT);
INSERT INTO test.simple(id, data) VALUES (1, '[1,2,3,4,5]');
```

#### Tests 

 - Unit test
 - Integration test
 - Manual test
 a. start tidb cluster 
 b. create a changefeed
 c. load vector data with https://github.com/breezewish/tidb-vector-bench
 d. check data consistency with sync_diff_inspector
  ```log
  A total of 1 tables need to be compared
  
  Comparing the table structure of `test`.`sample` ... equivalent
  Comparing the table data of `test`.`sample` ... equivalent
  _____________________________________________________________________________
  Progress [============================================================>] 100% 0/0
  A total of 1 table have been compared and all are equal.
  ```
##### Do you need to update user documentation, design documentation or monitoring documentation?
yes
### Release note 

```release-note
support vector data type
```

### CODE REVIEW

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 1]
```diff
@@ -348,8 +349,13 @@ func convertBinaryToString(cols []*model.ColumnData, tableInfo *model.TableInfo)
 			colValBytes, ok := col.Value.([]byte)
 			if ok {
 				cols[i].Value = string(colValBytes)
+				continue
 			}
 		}
+		switch v := col.Value.(type) {
```
Comment (CharlesCheung96) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:355]: Maybe the `[]byte` case should be handled here?

File: cdc/sink/ddlsink/mysql/async_ddl.go [hunk 2]
```diff
@@ -15,25 +15,29 @@ package mysql
...
 
 	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/dumpling/export"
 	timodel "github.com/pingcap/tidb/pkg/parser/model"
 	"github.com/pingcap/tiflow/cdc/model"
-	"github.com/pingcap/tiflow/pkg/errors"
 	"go.uber.org/zap"
 )
 
+const timeout = 5 * time.Second
```
Comment (3AceShowHand) [inline cdc/sink/ddlsink/mysql/async_ddl.go:28]: It looks this part changes not relevant to the vector type. It's suggested revert this part, and create a new PR to do this.

File: cdc/sink/ddlsink/mysql/format_ddl.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/pkg/parser"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/parser/format"
+	"github.com/pingcap/tidb/pkg/parser/mysql"
+	"go.uber.org/zap"
+)
+
+type visiter struct{}
```
Comment (3AceShowHand) [inline cdc/sink/ddlsink/mysql/format_ddl.go:27]: why we need this ?

File: cdc/sink/util/vector.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	"reflect"
+	"strconv"
+	"unsafe"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/pkg/types"
+	"go.uber.org/zap"
+)
+
+// CheckVectorIndexForColumnModify checks Vector Index constraints for MODIFY COLUMN.
```
Comment (3AceShowHand) [inline cdc/sink/util/vector.go:27]: if this is useless, please remove commented code.

File: cdc/sink/util/vector.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	if err := types.CheckVectorDimValid(dim); err != nil {
+		return types.ZeroVectorFloat32, err
+	}
+	vec := types.InitVectorFloat32(dim)
+	copy(vec.Elements(), values)
+	return vec, nil
+}
+
+// VectorElement2String parse elements to string
+func VectorElement2String(elements []interface{}) string {
```
Comment (3AceShowHand) [inline cdc/sink/util/vector.go:62]: What's the purpose of this method, looks so complicated, and error prone.

**Comment by 3AceShowHand** [comment]
why this is merged to [pingcap:feature/vector-data-type](https://github.com/pingcap/tiflow/tree/feature/vector-data-type)

File: cdc/sink/ddlsink/mysql/format_ddl.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		}
+	}
+	return n, false
+}
+
+func (f *visiter) Leave(n ast.Node) (node ast.Node, ok bool) {
+	return n, true
+}
+
+func formatQuery(sql string) string {
```
Comment (3AceShowHand) [inline cdc/sink/ddlsink/mysql/format_ddl.go:50]: only rewrite the DDL ?

File: cdc/sink/ddlsink/mysql/async_ddl.go (hunk 2 above)
Comment (wk989898) [reply cdc/sink/ddlsink/mysql/async_ddl.go:28]: It is related to https://github.com/pingcap/tiflow/issues/11521, I will revert it before merging.

File: cdc/sink/ddlsink/mysql/format_ddl.go (hunk 3 above)
Comment (wk989898) [reply cdc/sink/ddlsink/mysql/format_ddl.go:27]: It converts vector data type to other data type for some downstream.
```sql
CREATE table test.simple(id int primary key, data VECTOR(5)); # before
CREATE table test.simple(id int primary key, data LONGTEXT); # after
```

**Comment by wk989898** [comment]
> why this is merged to [pingcap:feature/vector-data-type](https://github.com/pingcap/tiflow/tree/feature/vector-data-type), instead of the master branch ?

 To keep the code tidy. Vector index support will also be provided in the future.

File: cdc/sink/ddlsink/mysql/format_ddl.go (hunk 6 above)
Comment (wk989898) [reply cdc/sink/ddlsink/mysql/format_ddl.go:50]: Yes, the vector format is similar to string in DML.

**Review by lance6716 — APPROVED** [review]

**Pull request merged**